packages/dds/merge-tree/src/segmentPropertiesManager.ts

```diff
@@ -45,133 +47,318 @@ export function copyPropertiesAndManager(
+		}
+		const rtn = applyChanges(op, seg, seq, (properties, deltas, key, value) => {
+			// eslint-disable-next-line unicorn/no-null
+			const previousValue = properties[key] ?? null;
 			if (collaborating) {
-				if (seq === UnassignedSequenceNumber) {
-					if (this.pendingKeyUpdateCount?.[key] === undefined) {
-						this.pendingKeyUpdateCount[key] = 0;
+				const pending: PropertyChanges | undefined = this.changes.get(key) ?? {
+					msnConsensus: previousValue,
+					remote: new DoublyLinkedList(),
+					local: new DoublyLinkedList(),
+				};
+				this.changes.set(key, pending);
+				const local = seq === UnassignedSequenceNumber;
+				if (local) {
+					pending.local.push(value);
+				} else {
+					// we only track remotes if there are adjusts, as only adjusts make application anti-commutative
+					// this will limit the impact of this change to only those using adjusts. Additionally, we only
+					// need to track remotes at all to support emitting the legacy snapshot format, which only sharedstring
+					// uses. when we remove the ability to emit that format, we can remove all remote op tracking
+					if (value.raw !== undefined && pending.remote.empty) {
+						pending.msnConsensus = computePropertyValue(pending.msnConsensus, [value]);
+					} else {
+						pending.remote.push(value);
 					}
-					this.pendingKeyUpdateCount[key]++;
-				} else if (!shouldModifyKey(key)) {
-					continue;
 				}
+				properties[key] = computePropertyValue(
+					pending.msnConsensus,
+					pending.remote.map((n) => n.data),
+					pending.local.map((n) => n.data),
+				);
+				if (local || pending.local.empty || properties[key] !== previousValue) {
+					deltas[key] = previousValue;
+				}
+			} else {
+				properties[key] = computePropertyValue(previousValue, [value]);
+				deltas[key] = previousValue;
 			}
+		});
+		this.updateMsn(msn);
+		return rtn;
+	}
 
-			const previousValue: unknown = oldProps[key];
-			// The delta should be null if undefined, as that's how we encode delete
-			// eslint-disable-next-line unicorn/no-null
-			deltas[key] = previousValue === undefined ? null : previousValue;
-			if (newValue === null) {
-				// eslint-disable-next-line @typescript-eslint/no-dynamic-delete
-				delete oldProps[key];
+	/**
+	 * Acknowledges property changes.
+	 * This method acknowledges the property changes based on the provided sequence number and operation.
+	 *
+	 * @param seq - The sequence number for the operation.
+	 * @param msn - The minimum sequence number for the operation.
+	 * @param op - The operation containing property changes.
+	 */
+	public ack(seq: number, msn: number, op: PropsOrAdjust): void {
+		for (const [key, value] of opToChanges(op, seq)) {
+			const change = this.changes.get(key);
+			const acked = change?.local?.shift();
+			assert(change !== undefined && acked !== undefined, "must have local change to ack");
+			// we only track remotes if there are adjusts, as only adjusts make application anti-commutative
```

Comment:
nice optimization, was pondering if the known issues we have around zamboni not being as active as it should would be problematic, but this should avoid that for the cases we care most about for now :)

Reply:
Yeah. i waffled on this a bit. I think it is fine for now, as i don't foresee these lists getting very big, and they only cost anything when changes are applied, and when there are changes, they get cleaned up. There might be some that sit around when they don't need to, but when we fix zamboni, this should also get resolved.